cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputCommitter.java

```diff
@@ -0,0 +1,243 @@
+        if (lastPathSepIdx == -1) {
+          // this shouldn't happen because each relative path should consist of at least one partition key and
+          // the output file name
+          LOG.warn("Skipping path '{}'. It's relative path '{}' has fewer than two parts", path, relativePath);
+          continue;
+        }
+        // relativePath = "../key1/key2/part-m-00000"
+        // relativeDir = "../key1/key2"
+        // fileName = "part-m-00000"
+        String relativeDir = relativePath.substring(0, lastPathSepIdx);
+        String fileName = relativePath.substring(lastPathSepIdx + 1);
+
+        Path finalDir = new Path(FileOutputFormat.getOutputPath(context), relativeDir);
+        Path finalPath = new Path(finalDir, fileName);
+        if (fs.exists(finalPath)) {
+          throw new FileAlreadyExistsException("Final output path " + finalPath + " already exists");
+        }
+        PartitionKey partitionKey = getPartitionKey(partitioning, relativeDir);
+        partitionsToAdd.add(partitionKey);
+        relativePaths.add(relativeDir);
+      }
+    }
+
+    // We need to copy to the parent of the FileOutputFormat's outputDir, since we added a _temporary_jobId suffix to
+    // the original outputDir.
+    Path finalOutput = FileOutputFormat.getOutputPath(context);
+    FileSystem fs = finalOutput.getFileSystem(context.getConfiguration());
+    for (FileStatus stat : getAllCommittedTaskPaths(context)) {
+      mergePaths(fs, stat, finalOutput);
+    }
+
+    // create all the necessary partitions
+    for (PartitionKey partitionKey : partitionsToAdd) {
+      outputDataset.getPartitionOutput(partitionKey).addPartition();
+    }
+
+    // close the TaskContext, which flushes dataset operations
+    try {
+      mrTaskContext.flushOperations();
+    } catch (Exception e) {
+      throw new IOException(e);
+    } finally {
+      mrTaskContext.close();
+    }
+
+    // delete the job-specific _temporary folder and create a _done file in the o/p folder
+    cleanupJob(context);
+
+    // mark all the final output paths with a _SUCCESS file, if configured to do so (default = true)
+    if (context.getConfiguration().getBoolean(SUCCESSFUL_JOB_OUTPUT_DIR_MARKER, true)) {
+      for (String relativePath : relativePaths) {
+        Path pathToMark = new Path(finalOutput, relativePath);
+        Path markerPath = new Path(pathToMark, SUCCEEDED_FILE_NAME);
+        fs.create(markerPath).close();
+      }
+    }
+  }
+
+  private PartitionKey getPartitionKey(Partitioning partitioning, String relativePath) {
+    List<String> pathParts = Lists.newArrayList(relativePath.split(Path.SEPARATOR));
+
+    PartitionKey.Builder builder = PartitionKey.builder();
+    int i = 0;
+    for (Map.Entry<String, Partitioning.FieldType> entry : partitioning.getFields().entrySet()) {
+      String keyName = entry.getKey();
+      Comparable keyValue = entry.getValue().parse(pathParts.get(i));
```

Comment:
Yea, the list is generated from the relative path of the files written (which is based upon the PartitionKey given by the user).
And the map comes from the partitioning of the dataset.
So as long as the PartitionKey that user returns is same size as the dataset's Partitioning, it'll be ok (except for the case that PartitionKey has '/' in one of its components).